File: Archipelago_ws_py/ap_client.py
import websockets
import asyncio
import json
import yaml
import re
import unicodedata
import urllib.request

with open('elden_ring.yaml', 'r') as f:
    config = yaml.safe_load(f)

server = config['server']
if not server.startswith(('ws://', 'wss://')):
    server = f'ws://{server}'


def clean_name(name: str) -> str:
    name = re.sub(r'^\[.*?\]\s*', '', name)
    name = unicodedata.normalize('NFKD', name).encode('ascii', 'ignore').decode('ascii')
    name = name.strip()
    if name == 'Terra Magicus':
        return 'Terra Magica'
    return name


class GameData:
    CACHE_FILE = 'datapackage_eldenring.json'

    def __init__(self):
        self._raw: dict = self._load_cache()
        self._flag_to_locs: dict[str, list[int]] | None = None  # cache interne

    def _load_cache(self) -> dict:
        try:
            with open(self.CACHE_FILE, 'r', encoding='utf-8') as f:
                return json.load(f)
        except FileNotFoundError:
            return {}

    def update(self, raw: dict):
        self._raw = raw
        self._flag_to_locs = None  # invalide le cache
        with open(self.CACHE_FILE, 'w', encoding='utf-8') as f:
            json.dump(raw, f, indent=4, ensure_ascii=False)
        print(f"DataPackage sauvegardé dans {self.CACHE_FILE}")

    @property
    def is_empty(self) -> bool:
        return not self._raw

    @property
    def _location_to_flag(self) -> dict:
        return (
            self._raw
            .get("data", {})
            .get("games", {})
            .get("EldenRing", {})
            .get("location_to_flag", {})
        )

    @property
    def flag_to_locs(self) -> dict[str, list[int]]:
        if self._flag_to_locs is None:
            result: dict[str, list[int]] = {}
            for loc_id, flag in self._location_to_flag.items():
                try:
                    result.setdefault(str(flag), []).append(int(loc_id))
                except (ValueError, TypeError) as e:
                    print(f"Warning: entrée invalide loc_id={loc_id} flag={flag}: {e}")
            self._flag_to_locs = result
        return self._flag_to_locs

    def watch_flags(self) -> list[int]:
        return [int(f) for f in self.flag_to_locs]

    def locs_for_flag(self, flag_id: int | str) -> list[int]:
        return self.flag_to_locs.get(str(flag_id), [])

    @property
    def item_name_to_id(self) -> dict[str, int]:
        return (
            self._raw
            .get("data", {})
            .get("games", {})
            .get("EldenRing", {})
            .get("item_name_to_id", {})
        )

    @property
    def item_id_to_name(self) -> dict[int, str]:
        return {v: k for k, v in self.item_name_to_id.items()}


with open('item_id_to_game_data.json', 'r', encoding='utf-8') as f:
    ITEM_GAME_DATA: dict[str, dict] = json.load(f)

_X_PAT = re.compile(r'^(.+) x(\d+)$')

secret = {"uuid": "afc006ee-e0f7-47b0-838b-7d2c37ed417b"}

queue_serv_to_dll = asyncio.Queue()
queue_dll_to_serv = asyncio.Queue()

game_data = GameData()


async def server_archi():
    received_idx = 0
    while True:
        try:
            async with websockets.connect(server) as websocket:
                raw = await websocket.recv()
                print(f"Received RoomInfo: {raw}")
                room_info = json.loads(raw)[0]
                server_version = room_info.get("version")

                if game_data.is_empty:
                    await websocket.send(json.dumps([{"cmd": "GetDataPackage", "games": ["EldenRing"]}]))
                    raw = await websocket.recv()
                    print(f"Received DataPackage: {raw[:200]}...")
                    game_data.update(json.loads(raw)[0])

                await websocket.send(json.dumps([{
                    "cmd": "Connect",
                    "password": config.get('password', ''),
                    "game": "EldenRing",
                    "name": config['name'],
                    "uuid": secret.get('uuid'),
                    "version": server_version,
                    "items_handling": 0b011,
                    "tags": [],
                    "slot_data": False
                }]))

                raw = await websocket.recv()
                response = json.loads(raw)[0]

                if response["cmd"] == "ConnectionRefused":
                    print(f"Refused: {response.get('errors')}")
                    return

                missing = response.get("missing_locations", [])
                checked = response.get("checked_locations", [])
                print(f"Connected! Team={response['team']} Slot={response['slot']}")
                print(f"Missing: {len(missing)} | Checked: {len(checked)}")

                async def send_loop():
                    while True:
                        msg = await queue_dll_to_serv.get()
                        print(f"[Server] Envoi vers AP: {msg}")
                        await websocket.send(json.dumps(msg))

                async def recv_loop():
                    nonlocal received_idx
                    id_to_name = game_data.item_id_to_name
                    async for message in websocket:
                        parsed = json.loads(message)
                        if isinstance(parsed, list) and parsed[0].get("cmd") == "ReceivedItems":
                            items = parsed[0].get("items", [])
                            index = parsed[0].get("index", 0)
                            print(f"[Server] ReceivedItems: index={index}, count={len(items)}")
                            for i, it in enumerate(items):
                                idx = index + i
                                if idx < received_idx:
                                    continue
                                item_id = it.get("item")
                                name = id_to_name.get(item_id)
                                if name is None:
                                    print(f"[Server]   [{idx}] item_id={item_id} inconnu, skip")
                                    continue
                                m = _X_PAT.match(name)
                                if m:
                                    base_name = m.group(1)
                                    qty = int(m.group(2))
                                else:
                                    base_name = name
                                    qty = 1
                                gd = ITEM_GAME_DATA.get(base_name)
                                if gd is None:
                                    print(f"[Server]   [{idx}] {name}: pas de game data, skip")
                                    continue
                                url = f"http://127.0.0.1:12999/give?base_id={gd['id']}&type={gd['type']}&qty={qty}"
                                try:
                                    urllib.request.urlopen(url, timeout=2)
                                    print(f"[Server]   [{idx}] {name} → /give OK (qty={qty})")
                                except Exception as e:
                                    print(f"[Server]   [{idx}] {name} → /give FAIL: {e}")
                                await asyncio.sleep(0.05)
                                received_idx = idx + 1
                        else:
                            print(f"[Server] Event AP: {parsed}")

                await asyncio.gather(send_loop(), recv_loop())

        except Exception as e:
            print(f"Connection error: {e}. Retry in 5s...")
            await asyncio.sleep(5)


async def dll():
    while True:
        try:
            async with websockets.connect("ws://127.0.0.1:12999/ws", ping_interval=20, ping_timeout=10) as websocket:
                print("[DLL] Connecté")

                flags = game_data.watch_flags()
                if flags:
                    loot_msg = {"set_flag_loot": flags}
                    print(f"[DLL] Envoi de {len(flags)} flags à surveiller")
                    await websocket.send(json.dumps(loot_msg))
                    try:
                        resp = await asyncio.wait_for(websocket.recv(), timeout=5.0)
                        print(f"[DLL] Réponse set_flag_loot: {resp}")
                    except asyncio.TimeoutError:
                        print("[DLL] Timeout set_flag_loot (continuant...)")

                async def recv_loop():
                    pending_locs: set[int] = set()

                    async def flush():
                        if pending_locs:
                            locs = list(pending_locs)
                            pending_locs.clear()
                            check_msg = [{"cmd": "LocationChecks", "locations": locs}]
                            print(f"[DLL] → LocationChecks batch: {len(locs)} locs")
                            await queue_dll_to_serv.put(check_msg)

                    async def reader():
                        async for message in websocket:
                            parsed = json.loads(message)
                            print(f"[DLL] Event: {parsed}")
                            if isinstance(parsed, dict) and parsed.get("type") == "flag_set":
                                if parsed.get("value") == 1:
                                    locs = game_data.locs_for_flag(parsed["flag_id"])
                                    print(f"[DLL] Flag {parsed['flag_id']} → locations: {locs}")
                                    pending_locs.update(locs)

                    async def flusher():
                        while True:
                            await asyncio.sleep(0.2)
                            await flush()

                    await asyncio.gather(reader(), flusher())

                await recv_loop()

        except Exception as e:
            print(f"[DLL] Error: {e}. Retry in 5s...")
            await asyncio.sleep(5)


async def main():
    await asyncio.gather(server_archi(), dll())

asyncio.run(main())

File: Rust_ws/src/ap_client.rs
use std::sync::Arc;
use std::time::Duration;

use anyhow::Result;
use futures_util::{SinkExt, StreamExt};
use tokio::sync::RwLock;
use tokio_tungstenite::{connect_async, tungstenite::Message};

use crate::config;
use crate::dll_client::DllMsg;
use crate::game_data::GameData;
use crate::given_locations::GivenLocations;

pub async fn run(
    config: config::Config,
    data_path: String,
    game_data: Arc<RwLock<GameData>>,
    dll_tx: tokio::sync::mpsc::UnboundedSender<DllMsg>,
    dll_rx: &mut tokio::sync::mpsc::UnboundedReceiver<DllMsg>,
    ap_dll_tx: tokio::sync::mpsc::UnboundedSender<crate::ApMsg>,
) {
    loop {
        let given_locs = GivenLocations::load(&data_path);
        let uuid = uuid::Uuid::new_v4().to_string();

        if let Err(e) = run_inner(&config, &game_data, &dll_tx, dll_rx, &ap_dll_tx, given_locs, &uuid).await {
            eprintln!("[AP] Error: {e}. Retry in 5s...");
        }
        tokio::time::sleep(Duration::from_secs(5)).await;
    }
}

async fn run_inner(
    config: &config::Config,
    game_data: &Arc<RwLock<GameData>>,
    _dll_tx: &tokio::sync::mpsc::UnboundedSender<DllMsg>,
    dll_rx: &mut tokio::sync::mpsc::UnboundedReceiver<DllMsg>,
    ap_dll_tx: &tokio::sync::mpsc::UnboundedSender<crate::ApMsg>,
    mut given_locs: GivenLocations,
    uuid: &str,
) -> Result<()> {
    let url = config::ap_url(config);
    let (ws_stream, _) = connect_async(&url).await?;
    let (mut write, mut read) = ws_stream.split();

    println!("[AP] Connecté au serveur {}", url);

    // 1. Receive RoomInfo
    let room_info = match read.next().await {
        Some(Ok(Message::Text(text))) => {
            let parsed: Vec<serde_json::Value> = serde_json::from_str(&text)?;
            parsed.into_iter().next().unwrap_or_default()
        }
        _ => return Err(anyhow::anyhow!("Pas de RoomInfo reçu")),
    };

    println!("[AP] RoomInfo reçu");
    let server_version = room_info["version"].clone();
    let seed_name = room_info["seed_name"].as_str().unwrap_or("").to_string();
    let server_checksum = room_info["datapackage_checksums"]["EldenRing"]
        .as_str()
        .map(|s| s.to_string());

    given_locs.set_seed(&seed_name);

    // 2. Check/update datapackage
    let needs_update = {
        let gd = game_data.read().await;
        gd.datapackage.is_none() || gd.checksum() != server_checksum
    };

    if needs_update {
        println!("[AP] Demande du DataPackage...");
        let req = r#"[{"cmd":"GetDataPackage","games":["EldenRing"]}]"#;
        write.send(Message::Text(req.into())).await?;

        if let Some(Ok(Message::Text(text))) = read.next().await {
            let parsed: Vec<serde_json::Value> = serde_json::from_str(&text)?;
            if let Some(dp) = parsed.into_iter().next() {
                let mut gd = game_data.write().await;
                gd.update_datapackage(dp);
                println!("[AP] DataPackage mis à jour");
            }
        }
    }

    // 3. Send Connect
    let items_handling = config.items_handling.unwrap_or(3);
    let connect_msg = serde_json::json!([{
        "cmd": "Connect",
        "password": config.password,
        "game": "EldenRing",
        "name": config.name,
        "uuid": uuid,
        "version": server_version,
        "items_handling": items_handling,
        "tags": [],
        "slot_data": false,
    }]);
    write.send(Message::Text(connect_msg.to_string().into())).await?;
    println!("[AP] Connect envoyé");

    // 4. Receive Connected
    let connected = match read.next().await {
        Some(Ok(Message::Text(text))) => {
            let parsed: Vec<serde_json::Value> = serde_json::from_str(&text)?;
            parsed.into_iter().next().unwrap_or_default()
        }
        _ => return Err(anyhow::anyhow!("Pas de réponse à Connect")),
    };

    if connected["cmd"] == "ConnectionRefused" {
        let errors = connected.get("errors");
        eprintln!("[AP] Connexion refusée: {:?}", errors);
        return Err(anyhow::anyhow!("ConnectionRefused"));
    }

    let team = connected["team"].as_u64().unwrap_or(0);
    let slot = connected["slot"].as_u64().unwrap_or(0);
    println!("[AP] Connecté! Team={team} Slot={slot}");

    // 5. Main event loop
    let mut received_idx: u64 = 0;

    loop {
        tokio::select! {
            Some(msg) = dll_rx.recv() => {
                let json = match msg {
                    DllMsg::LocationChecks(locs) => {
                        serde_json::json!([{"cmd": "LocationChecks", "locations": locs}])
                    }
                    DllMsg::StatusUpdate(status) => {
                        serde_json::json!([{"cmd": "StatusUpdate", "status": status}])
                    }
                };
                write.send(Message::Text(json.to_string().into())).await?;
            }
            ws_msg = read.next() => {
                match ws_msg {
                    Some(Ok(Message::Text(text))) => {
                        let parsed: Vec<serde_json::Value> = serde_json::from_str(&text)?;
                        for msg in parsed {
                            let cmd = msg["cmd"].as_str().unwrap_or("");
                            match cmd {
                                "ReceivedItems" => {
                                    let items = msg["items"].as_array().cloned().unwrap_or_default();
                                    let index = msg["index"].as_u64().unwrap_or(0);

                                    for (i, item) in items.iter().enumerate() {
                                        let idx = index + i as u64;
                                        if idx < received_idx {
                                            continue;
                                        }
                                        let item_id = match item["item"].as_u64() {
                                            Some(id) => id as u32,
                                            None => continue,
                                        };
                                        let location_id = match item["location"].as_u64() {
                                            Some(id) => id as u32,
                                            None => continue,
                                        };

                                        if given_locs.has(location_id) {
                                            continue;
                                        }

                                        let name = {
                                            let gd = game_data.read().await;
                                            gd.item_id_to_name(item_id).map(|s| s.to_string())
                                        };

                                        if let Some(name) = name {
                                            let (base_name, qty) = crate::dll_client::parse_item_name(&name);
                                            let gd = {
                                                let gd = game_data.read().await;
                                                gd.item_game_data(base_name).cloned()
                                            };
                                            if let Some(item_data) = gd {
                                                let msg = crate::ApMsg {
                                                    base_id: item_data.id,
                                                    item_type: item_data.item_type,
                                                    qty,
                                                };
                                                ap_dll_tx.send(msg)?;
                                                given_locs.add(location_id);
                                                println!("[AP] [{idx}] {name} → give (qty={qty})");
                                            } else {
                                                println!("[AP] [{idx}] {name}: pas de game data, skip");
                                            }
                                        } else {
                                            println!("[AP] [{idx}] item_id={item_id} inconnu, skip");
                                        }
                                        received_idx = idx + 1;
                                    }
                                }
                                "RoomUpdate" => {
                                    let missing = msg["missing_locations"].as_array().map(|a| a.len()).unwrap_or(0);
                                    let checked = msg["checked_locations"].as_array().map(|a| a.len()).unwrap_or(0);
                                    println!("[AP] RoomUpdate: missing={missing} checked={checked}");
                                }
                                "PrintJSON" => {
                                    if let Some(data) = msg.get("data") {
                                        println!("[AP] PrintJSON: {data}");
                                    }
                                }
                                _ => {
                                    if !cmd.is_empty() {
                                        println!("[AP] Event: cmd={cmd}");
                                    }
                                }
                            }
                        }
                    }
                    Some(Ok(Message::Close(_))) | None => {
                        println!("[AP] Déconnecté du serveur");
                        break;
                    }
                    _ => {}
                }
            }
        }
    }

    Ok(())
}
